File: DataStructures/HashSet.py

class HashSet1:

    def __init__(self, num_buckets=10):

        self.buckets = [ None ] * num_buckets

    def _hash(self, item):

        return item % len(self.buckets)

    def add(self, item):

        index = self._hash(item)

        if self.buckets[index] == None:

            self.buckets[index] = [item]

        elif item not in self.buckets[index]:

            self.buckets[index].append(item)

    def remove(self, item):

        index = self._hash(item)

        self.buckets[index].remove(item)

    def __contains__(self, item):

        index = self._hash(item)

        return item in self.buckets[index]

class HashSet2:

    def __init__(self):

        self.hashset = set() # set() is O(1)

    def add(self, item):

        self.hashset.add(item)

    def remove(self, item):

        self.hashset.discard(item)

    def __contains__(self, item):

        return item in self.hashset

def main():

    for hashset_class in [HashSet1, HashSet2]:

        hashset = hashset_class()

        hashset.add(1)
        hashset.add(2)
        hashset.add(1)

        print(1 in hashset)

        hashset.remove(1)

        print(1 in hashset)

        print(2 in hashset)

        print("------")

if __name__ == '__main__':
    main()

File: DataStructures/HashSet.js
"use strict";

class HashSet1 {

    constructor(numBuckets = 10) {

        this.buckets = [];
        while (this.buckets.push([]) < numBuckets);

    }

    _hash(item) {

        return item % this.buckets.length;

    }

    add(item) {

        let index = this._hash(item);

        if (!this.buckets[index].includes(item)) {

            this.buckets[index].push(item);

        }

    }

    has(item) {

        let index = this._hash(item);

        return this.buckets[index].includes(item);

    }

    remove(item) {

        let index = this._hash(item);

        this.buckets[index].splice(this.buckets[index].indexOf(item), 1);

    }

}

class HashSet2 {

    constructor(buckets = 10) {

        this.hashTable = {}; // Object pairs are implemented as a hash table

    }

    add(item) {

        this.hashTable[item] = true;

    }

    has(item) {

        return this.hashTable[item] || false;

    }

    remove(item) {

        this.hashTable[item] = false;

    }

}

for (let hashSetClass of [HashSet1, HashSet2]) {

    let hashSet = new hashSetClass();

    hashSet.add(1);
    hashSet.add(2);
    hashSet.add(1);

    console.log(hashSet.has(1));

    hashSet.remove(1);

    console.log(hashSet.has(1));

    console.log(hashSet.has(2));

    console.log("------");

}
